validator_client/validator_services/src/head_monitor_service.rs

```diff
@@ -0,0 +1,223 @@
+#[derive(Debug)]
+pub struct HeadEvent {
+    pub beacon_node_index: usize,
+}
+
+impl<S: ValidatorStore + 'static, T: SlotClock + 'static> HeadMonitorServiceBuilder<S, T> {
+    pub fn new() -> Self {
+        Self {
+            validator_store: None,
+            beacon_nodes: None,
+            executor: None,
+            beacon_head_cache: None,
+            head_monitor_tx: None,
+        }
+    }
+
+    pub fn beacon_nodes(mut self, nodes: Arc<BeaconNodeFallback<T>>) -> Self {
+        self.beacon_nodes = Some(nodes);
+        self
+    }
+
+    pub fn executor(mut self, executor: TaskExecutor) -> Self {
+        self.executor = Some(executor);
+        self
+    }
+
+    pub fn validator_store(mut self, validator_store: Arc<S>) -> Self {
+        self.validator_store = Some(validator_store);
+        self
+    }
+
+    pub fn beacon_head_cache(mut self, beacon_head_cache: Arc<BeaconHeadCache>) -> Self {
+        self.beacon_head_cache = Some(beacon_head_cache);
+        self
+    }
+
+    pub fn head_monitor_tx(mut self, head_monitor_tx: Arc<mpsc::Sender<HeadEvent>>) -> Self {
+        self.head_monitor_tx = Some(head_monitor_tx);
+        self
+    }
+    pub fn build(self) -> Result<HeadMonitorService<S, T>, String> {
+        let beacon_head_cache = Arc::new(BeaconHeadCache {
+            cache: RwLock::new(HashMap::new()),
+        });
+        Ok(HeadMonitorService {
+            inner: Arc::new(Inner {
+                _validator_store: self
+                    .validator_store
+                    .ok_or("Cannot build HeadMonitorService without validator_store")?,
+                beacon_nodes: self
+                    .beacon_nodes
+                    .ok_or("Cannot build HeadMonitorService without beacon_nodes")?,
+                executor: self
+                    .executor
+                    .ok_or("Cannot build HeadMonitorService without executor")?,
+                beacon_head_cache,
+                head_monitor_tx: self
+                    .head_monitor_tx
+                    .ok_or("Cannot build HeadMonitorService without head_monitor_rx")?,
+            }),
+        })
+    }
+}
+
+// Helper to minimise `Arc` usage
+pub struct Inner<S, T> {
+    _validator_store: Arc<S>,
+    executor: TaskExecutor,
+    beacon_nodes: Arc<BeaconNodeFallback<T>>,
+    beacon_head_cache: Arc<BeaconHeadCache>,
+    head_monitor_tx: Arc<mpsc::Sender<HeadEvent>>,
+}
+
+// Runs a non-terminating loop to update the `BeaconHeadCache` with the latest head received
+// from the candidate beacon_nodes. This is an attempt to stream events to beacon nodes and
+// potential start attestion duties earlier as soon as latest head is receive from any of the
+// beacon node in contrast to attest at the 1/3rd mark in the slot.
+//
+// The cache and the candidate BNs list are periodically refresh/purged to avoid race condition that may
+// be arise due to change in ranking of beacon_nodes thus affecting the indices of beacon_nodes
```

Comment:
I believe ranking beacon_nodes doesn't touch their internal index (`candidate.index`). It just affects their actual ordering in the vec, which only really matters for `first_success`.